### DISCUSSION TITLE
How to use `.env` file with Quadlet `.kube`?

### DISCUSSION THREAD

**Discussion by poperigby** [containers/podman#21184] [Q&A]
Does `podman kube play <service>.yaml` automatically look for `.env` files in the directory of the YAML file, or do I have to manually specify them?

**Comment by rhatdan** [comment]
Is .env a kubernetes concept?
@ygalblum PTAL

**Reply by ygalblum** [reply]
AFAIK this is a `docker-compose` concept: https://docs.docker.com/compose/environment-variables/set-environment-variables/#substitute-with-an-env-file

**Reply by rhatdan** [reply]
@poperigby so this is not supported, what is your actual goal?

**Reply by poperigby** [reply]
Oh, I see. I wanted somewhere to place secrets that don't get committed to my git repository. I could use Podman secrets, but I'm not sure how to use a Podman secret created through the CLI in a Kubernetes YAML.

**Reply by ygalblum** [reply]
In order to use a secret in a K8S YAML, the secret itself should also be a K8S secret.
You can see the `Create secrets` section in https://www.redhat.com/sysadmin/multi-container-application-podman-quadlet.
If you don't want to use `kubectl` you can also prepare your secret as a jinja template. See the `Secrets` section in https://inlets.dev/blog/2023/10/03/client-quadlet.html.

**Reply by poperigby** [reply]
Thank you. I think neither of those options are ideal. The first one because then I would be committing secrets to my git repository, and the second one because it seems like quite a bit of work, and an inelegant way to have secrets, especially if I have to do it for every Kube YAML. Podman being able to read from a .env file, or using Podman CLI created secrets would be much preferred.

**Reply by rhatdan** [reply]
Secrets are secret to the application that is running, and the committed image ,not necessarily to the user running the containers.

**Reply by poperigby** [reply]
Is there any feature that would allow me to keep sensitive information separate from the configuration of containers? If not, that would be useful.

**Reply by rhatdan** [reply]
You can use environment variables via systemd if you want, or environment variables set in a quadlet. Or use podman secrets.
Not sure how this is much different?

**Reply by poperigby** [reply]
Is it possible to then use those in my YAML file?

**Reply by rhatdan** [reply]
I believe you can use the secrets in the YAML file as environment variables. @ygalblum @umohnani8 Thoughts?

**Reply by ygalblum** [reply]
As I wrote earlier, in order to use a secret in a K8S Pod, the Podman secret itself must also be in K8S Secret. As noted in the `Create secrets` section of my [blog](https://www.redhat.com/sysadmin/multi-container-application-podman-quadlet), since the demo uses both `.container` (which translated to a `podman run` command) and `.kube` (which translated to a `podman kube play` command) files, the same information (`MYSQL_ROOT_PASSWORD`) is stored in two secrets.

The reason is that when using a secret with `podman run` the entire content of the secret is uses, while when using a secret with `podman kube play` Podman parses the content of the secret as a K8S Secret and only the `data` field (and the keys within it) is used.

Currently, I don't think there is an automatic way to combine separated configuration and data (in this case secret) and I understand the request.

However, we need to keep in mind that `podman kube play` tries to stay aligned with the behavior of `kubectl`. Both support  substitution based on the environment variables defined in the Pod spec (see [here](https://kubernetes.io/docs/tasks/inject-data-application/define-environment-variable-container/#using-environment-variables-inside-of-your-config)). So, how will such an automatic process know whether a placeholder should be substituted with Host machine variables or with the ones defined in the Pod Spec?

Not sure how to address this
